internal/controller/controllers/installenv_controller.go

```diff
@@ -0,0 +1,221 @@
+	"github.com/openshift/assisted-service/internal/common"
+	adiiov1alpha1 "github.com/openshift/assisted-service/internal/controller/api/v1alpha1"
+	"github.com/openshift/assisted-service/models"
+	"github.com/openshift/assisted-service/restapi/operations/installer"
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	ImageCreated       = "ImageCreated"
+	ImageCreationError = "ImageCreationError"
+)
+
+type ImageState struct {
+	Status    corev1.ConditionStatus
+	State     string
+	StateInfo string
+}
+
+// InstallEnvReconciler reconciles a InstallEnv object
+type InstallEnvReconciler struct {
+	client.Client
+	Log                      logrus.FieldLogger
+	Installer                bminventory.InstallerInternals
+	PullSecretUpdatesChannel chan event.GenericEvent
+}
+
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=installenvs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=installenvs/status,verbs=get;update;patch
+
+func (r *InstallEnvReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	installEnv := &adiiov1alpha1.InstallEnv{}
+	if err := r.Get(ctx, req.NamespacedName, installEnv); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	return r.ensureISO(ctx, installEnv)
+}
+
+// ensureISO generates ISO for the cluster if needed and will update the condition Reason and Message accordingly.
+// It returns a result that includes ISODownloadURL.
+func (r *InstallEnvReconciler) ensureISO(ctx context.Context, installEnv *adiiov1alpha1.InstallEnv) (ctrl.Result, error) {
+	var inventoryErr error
+	var imageInfo *models.ImageInfo
+	var updatedCluster *common.Cluster
+	imageState := ImageState{
+		Status:    corev1.ConditionUnknown,
+		State:     ImageCreationError,
+		StateInfo: adiiov1alpha1.ImageStateFailedToCreate,
+	}
+
+	kubeKey := types.NamespacedName{
+		Name:      installEnv.Spec.ClusterRef.Name,
+		Namespace: installEnv.Spec.ClusterRef.Namespace,
+	}
+	clusterDeployment := &hivev1.ClusterDeployment{}
+
+	if err := r.Get(ctx, kubeKey, clusterDeployment); err != nil {
+		clusterDeploymentRefErr := newKubeAPIError(
+			errors.Wrapf(
+				err,
+				fmt.Sprintf("failed to find clusterDeployment with name %s in namespace %s",
+					installEnv.Spec.ClusterRef.Name, installEnv.Spec.ClusterRef.Namespace)),
+			k8serrors.IsNotFound(err),
+		)
+		return r.updateStatusAndReturnResult(ctx, installEnv, nil, imageState, clusterDeploymentRefErr)
```

Comment:
My personal opinion is that the whole flow would be easier to follow if we just set the condition directly instead of using `ImageState` and passing it to `updateStatusAndReturnResult`.

```suggestion
		conditions.SetStatusCondition(&installEnv.Status.Conditions, conditions.Condition{
				Type:   adiiov1alpha1.ImageCreatedCondition,
				Status: corev1.ConditionUnknown,
				Reason: "ImageCreationError",
				Message: adiiov1alpha1.ImageStateFailedToCreate,
		})
		return ctrl.Result{Requeue: true}, nil
```

Of course, this would require changes in `internal/controller/api/v1alpha1/installenv_types.go` to make what is now a `InstallEnvConditionType` be compatible with the `openshift/custom-resource-status` lib.

Reply:
Done